### DISCUSSION TITLE
Send callback response back to initial request

### DISCUSSION THREAD

**Discussion by inuitviking** [mockoon/mockoon#1561] [Help] [answered]
Hi!

I love this tool, and have been using it a lot for work; quite exceptional.

When creating an HTTP route, and setting it to use a callback, it would be cool to have it return the response from the callback. I don't seem to be able to do that. If I have missed anything, please let me know!

For example, when creating a new HTTP Route, instead of having the body being Inline, File, and Data only, we could have Callback Response body. If this is achievable by templates instead, I would like to know how (still learning!).

Hope this issue finds you well. :)

**Comment by inuitviking** [comment]
Since it's a discussion now, it would also be cool to do more with callbacks in general, especially with the potential response body.

I unfortunately don't know much about JS to make a pull request, but I'm willing to test.

**Comment by 255kb** [comment]
Hi

> I love this tool, and have been using it a lot for work; quite exceptional.

Thank you for the kind words!

I'm not sure this will be possible anytime soon. The callbacks are separated from the request, eventually delayed. The response is usually already sent when the callback's response arrives. We could imagine a system where we store callback's responses in memory, but using them immediately in the initial request/response seems to contradict how callback/webhook traditionally work imo. (a request is sent, we receive a response, later the server trigger a callback, but there is no link between the two transactions).

**Reply by inuitviking** [reply]
The reason I thought that this would be cool, would be that I could mock a not-very-RESTful backend.

I'm trying to mock a backend that has endpoints like these:

- `/transactions`
- `/transaction`
- `/transaction/clear`

Whereas with a RESTful API, there would only be one endpoint handling everything about transactions. All of this is subject to change in the not-so-very-close future.

With callbacks, and imagine you could get its response, I could create a CRUD route, and then the various non-REST compliant HTTP routes, with the CRUD route as their callback.

I have a way to make this *almost* work with templating, but for that I kind of need help. I'll get back to this once I'm at work.

**Reply by inuitviking** [reply]
I have this setup at work:

- A databucket called "transactions"
  - It's a JSON array containing JSON objects, with a field called `transactionId`
- A CRUD-route called `/transactions-crud`
- Three HTTP-routes:
  - `/transactions/status`
    - Gets the entire transactions array
  - `/transaction`
    - The `transactionId` is passed via the body
  - `/transaction/clear`
    - The `transactionId` is passed via the body

The `/transaction/clear` HTTP-route has a callback (`DELETE`) to the `/transactions/{{body 'transactionId'}}` endpoint. This one is easy, as I simply template the response like so:

```json
{
  "transactionId": "{{body 'transactionId'}}",
  "cleared": true
}
```

It becomes more tricky when I want to just fetch a single transaction from transactions. If I define the ID myself in the template, like so, it works:
```
{{{ stringify ( filter (dataRaw 'transactions') (object siteTransactionId='1234') ) }}}
```

But this unfortunately doesn't:
```
{{{ stringify ( filter (dataRaw 'transactions') (object siteTransactionId="(body 'transactionId')") ) }}}
```

This lead me to the weird thought that getting the callback response would be cool, as that would be, in my case at least, an ordinary CRUD-route, which is easier to work with. :sweat_smile:

**Reply by 255kb** [reply, marked as answer]
I see what you are trying to do! I was recently facing not-so-RESTful APIs and indeed, the CRUD routes are sometimes too constrained (but they cannot accommodate every use cases). I also still think the callbacks are a bit out of their scope here. 

I don't know if you had the time to look at [v9](https://mockoon.com/releases/9.0.0/) which was released last week. 
It includes a new [`setData` helper](https://mockoon.com/docs/latest/templating/mockoon-helpers/#setdata) that can manipulate a data bucket with operators like `set`, `push`, etc.
With this helper you should be able to create your 3 or 4 routes to manipulate a bucket like you would with a REST API, without the need for callbacks. 

Regarding the filter, I think the correct syntax is this one, without the quotes around `(body 'transactionId')`: 

```handlebars
{{{ stringify ( filter (dataRaw 'transactions') (object siteTransactionId=(body 'transactionId') ) ) }}}
```

**Reply by inuitviking** [reply]
Thanks for the reply! I  haven't had time to look at v9, as we're nearing launch at work, but I supposed in the next few weeks, I may have the time. I'll mark your comment as the answer, as it seems to be what we need; otherwise, I'll just reopen. :smile:

**Reply by inuitviking** [reply]
@255kb Hello again!

I got to play with this sooner than I expected. Your help lead me in the right direction, and I have a (almost) un-fully RESTful API mock setup.

One thing is bothering me, though, which I can't seem to get my head around.

```
{{ setData 'push' 'pumps' ( concat '$[?(@.pumpId == ' ( urlParam 'pumpId' ) "].pumpStateDesc"  ) 'Fuelling' }}
```

I have this thing where I have to change a pumps status description to "Fuelling", and it has to be the specific pump passed from the URL parameter. The issue for me seems to be that when I use concat, the `==` are converted to `&#x3D;&#x3D;`.

Is there another way to pass the urlParam to `setData`?

**Reply by inuitviking** [reply]
I have found something that works for me, but doesn't cover all situations:

```
{{ setVar 'pumpId' (subtract (parseInt (urlParam 'pumpId')) 1) }}
{{ setData 'set' 'pumps' ( concat @pumpId '.pumpStateDesc' ) 'Fuelling' }}
{{ setData 'set' 'pumps' ( concat @pumpId '.terminalStatus' )  '1' }}
{{{ stringify ( oneOf ( filter ( dataRaw 'pumps' ) ( object pumpId=( parseInt ( urlParam 'pumpId') ) ) ) ) }}}
```

It works for me, because the pumpId is incremental, and I start from "1", so get the urlParam, subtract by one, use object-path instead of the other tool, and concat the pumpId to the key you want to change.

Would be cool to be able to use JSONPath, as it can query, but currently, the escaping of equal characters is not something I have figured out yet.

**Reply by inuitviking** [reply]
Replying again, because I ran into another place where I couldn't just rely on the index, but needed a specific UUID.
The databucket looks something like this (transactionId is simplified to e.g. "1234"):

```json
[
  {
    "pumpenr": 1,
    "transactionId": "1234",
    "produktnr": 13,
    "valuta": "DKK",
    "beloeb": "100.75",
    "liter": "100.75",
    "apris": 1
  },
  {
    "pumpenr": 1,
    "transactionId": "4321",
    "produktnr": 13,
    "valuta": "DKK",
    "beloeb": "100.75",
    "liter": "100.75",
    "apris": 1
  }
]
```

Here's how I solved it:

```mockoon
{{ setGlobalVar 'counter' 0 }}
{{ setGlobalVar 'cleared' false }}
{{#each ( dataRaw 'transactions' )}}
  {{{ setVar 'transactionId' ( stringify this.transactionId ) }}}
  {{{ setVar 'releaseToken' ( stringify ( bodyRaw 'releaseToken' ) ) }}}
  {{{ setVar 'pumpIndex' ( subtract ( parseInt ( stringify this.pumpenr ) ) 1 ) }}}
  {{#if ( eq ( stringify this.transactionId ) (stringify ( bodyRaw 'releaseToken' ) ) ) }}
    {{ setData 'del' 'transactions' @counter }}
    {{ setData 'set' 'pumps' ( concat @pumpIndex '.pumpStateDesc' ) 'Idle' }}
    {{ setData 'set' 'pumps' ( concat @pumpIndex '.terminalStatus' )  '0' }}
    {{ setGlobalVar 'cleared' true }}
  {{/if}}
  {{ setGlobalVar 'counter' (add (getGlobalVar 'counter') 1) }}
{{/each}}
{
  "cleared": {{getGlobalVar 'cleared'}},
  "counter": {{getGlobalVar 'counter'}}
}
```

In other words:

- Create a global variable and set the value to 0 - This is your index key and will be incremented
- Create a global variable to tell whether a transaction was cleared (default false, true if deleted)
- Loop through all transactions
- Create three variables, one containing the UUID from the object in the databucket, one from the body in the POST request, one to hold the pumpIndex (all three are stringified, to make them comparable and easier for lookup)
- Compare transactionId and releaseToken and check if they're equal
- If they are:
  - Delete the transaction at index `counter`
  - Set the pump at pumpIndex to have pumpStateDesc as "Idle" and "terminalStatus" as 0
  - Set the global variable "cleared" to true
- Increment `counter` by 1
- Print out a JSONObject containing both `cleared` and `counter`.

The reason "pumpenr" is the same, is because there can be multiple transactions on the same pump, that's why we need to look at "transactionId". This seems to do an excellent job. :)

Hope this helps whomever may run into the same issues as I have.

**Reply by 255kb** [reply]
Hi, glad you found a solution! 
You first example cannot work as the `setData` helper doesn't support JSONPath, only object-path (`prop.0.prop`). The reason is that  the object-path lib we are using can get and set properties out of the box, while JSONPath can only get. 
What you did with `each` is probably an easier approach.

Regarding the escaping of characters it shouldn't happen in the concat + setData helpers. You may see it if you try something like `{{concat '$.[?(@.pumpId == ' ( urlParam 'pumpId' ) '].pumpStateDesc' }}` as Handlebars will escape the characters. You can avoir this by using the triple curly braces, but anyway it shouldn't happen when used with the other helpers (the reason is rather that JSONPath is not supported in the `setData` helper).

**Reply by inuitviking** [reply]
Ah, that explains it. Thanks for clarifying!

Using `each` and object-path in `setData` does work wonders, though, so I'm happy I can do exactly what I need to do.

Thanks for this spectacular tool. I can't get enough of it. :smiling_face_with_three_hearts:

**Reply by 255kb** [reply]
You're welcome! Do not hesitate if there are small improvements we could make to the helpers. Sometimes, we need to do really circumvoluted things where one change or one new helper would help a lot. (Also I have no issue adding lots of helpers as they don't take any UI estate 😀 )

**Reply by inuitviking** [reply]
I'll start a new discussion or feature request if I come up with anything!
